Add ExecuteScript feature to launch Google APIs

athevret opened the pull request:
Hello [Dariusz](https://github.com/darrarski),
I used your framework for my application [swatDocuments](https://apple.co/41jLWHq) and I needed to execute a Google Apps Script in the same Google session as the Google Drive. I made a fork to [GoogleDriveClient](https://github.com/darrarski/swift-google-drive-client) and added a new feature to execute scripts and integrated it in my fork. If you want to add it, I would be more than happy.  The are some aspects of your developed solution with "dependencies" that I'm not expert in, and this is the reason why I did not add tests for it.
Have a nice day. Let me know if you have questions.
[Alban](https://github.com/athevret)

Commit:
Add ExecuteScript feature to launch Google APIs

darrarski commented on Sources/GoogleDriveClient/ExecuteScript.swift:
```diff
@@ -0,0 +1,106 @@
+        components.scheme = "https"
+        components.host = "script.googleapis.com"
+        components.path = "/v1/scripts/\(params.scriptId):run"
+        print("Components: \(components)")
```
Remove print statement.

athevret replied:
Agree. Removed.

darrarski commented on Sources/GoogleDriveClient/ExecuteScript.swift:
```diff
@@ -0,0 +1,106 @@
+          "\(credentials.tokenType) \(credentials.accessToken)",
+          forHTTPHeaderField: "Authorization"
+        )
+        print("Authorization: \(credentials.tokenType) \(credentials.accessToken)")
```
Remove print statement.

athevret replied:
Agree. removed.

darrarski commented on Sources/GoogleDriveClient/ExecuteScript.swift:
```diff
@@ -0,0 +1,106 @@
+          "parameters": params.args,
+          "devMode": false
+        ]
+        request.httpBody = try? JSONSerialization.data(withJSONObject: requestBody)
```
We should not ignore the error with `try?`. If the request body can't be encoded, we should throw the error.

```swift
let request: URLRequest = try {
  ...
  request.httpBody = try JSONSerialization.data(withJSONObject: requestBody)
  ...
}
```

athevret replied:
Agree. Removed "?" to propagate error.

darrarski reviewed (commented):
Looks good, thanks for your contribution! I've added some comments that should be addressed before I merge it. 

I'm not familiar with Apps Script API, but I see that Google [provides documentation](https://developers.google.com/apps-script/api/reference/rest/v1/scripts/run#response-body) for the response body. Please consider adding a Swift struct, like [`ExecutionResponse`](https://developers.google.com/apps-script/api/reference/rest/v1/ExecutionResponse) and use it as a result of `ExecuteScript.Run` (instead of returning `Data`). As an example, take a look at `GetFile` interface that returns `File` struct. 

I'm afraid the current way of how this is implemented is not entirely correct, as Google API can return error info in the response body. This should be handled by the `ExecuteScript` and an error should thrown in such a case.

I see you followed my approach to designing dependencies without a problem. The only thing that's missing is a test case for the life implementation of `ExecuteScript`.

There are test cases for other endpoints already in the repository. For example, take a look at these files:

- `Tests/GoogleDriveClientTests/GetFileTests.swift`
- `Tests/GoogleDriveClientTests/GetAboutTests.swift`
- and others

I will be happy to review the tests for `ExecuteScript.live` once you add it. If you have any questions - feel free to ask.

darrarski commented on Sources/GoogleDriveClient/ExecuteScript.swift:
```diff
@@ -0,0 +1,106 @@
+        components.path = "/v1/scripts/\(params.scriptId):run"
+        print("Components: \(components)")
+
+        var request = URLRequest(url: components.url!)
```
It's (mostly) safe to force unwrap here, as long as we cover it with unit tests.

In theory, there is a chance that url can't be constructed here, if for some reason the script id (included in the path) makes the URL invalid - this will crash. To be 100% sure the code is safe, no matter the input parameters - it's better to catch it here and throw an appropriate error instead of force unwrapping (like `Error.invalidParams`).

athevret replied:
Added invalidParams error and raised that error in case url is not valid.
But actually, during unit tests, I found impossible to raise that error. All incorrect characters are escaped with percent encoding. I left the error raised and commented out the unit test to keep track of the effort. If you have an idea of how to raise the error, you can uncomment and change the test.

darrarski commented on Sources/GoogleDriveClient/ExecuteScript.swift:
```diff
@@ -0,0 +1,106 @@
+          "parameters": params.args,
+          "devMode": false
+        ]
+        request.httpBody = try? JSONSerialization.data(withJSONObject: requestBody)
```
It's not necessary to wrap the errors thrown from `JSONSerialization` into custom error type (`Error.serializationError(underlyingError:)`). Throwing the error instead of ignoring it should be good enough - like described in https://github.com/darrarski/swift-google-drive-client/pull/26#discussion_r2083359444.

athevret replied:
Agree. Removed "?" to propagate error.

Commit:
Fixed comments from "Add ExecuteScript feature to launch Google APIs" MR #26
- removed print statements,
- added unit tests,
- raised exception for invalid parameters.

athevret commented:
I updated my fork with a set of unit tests and tested with my swatDocuments app the modified error management (now performed partially in the ExecuteScript API). Hopefully you will receive the update in the MR. Let me know if you wan additional changes. Best regards. Alban
